config/7.2.0/securebanking/ig/scripts/groovy/SignEventsResponse.groovy

```diff
@@ -0,0 +1,102 @@
+    }
+}
+
+Error response example:
+{
+    "Code": "500",
+    "Id": "ede36b552dc951d9836a127f16a7c033",
+    "Message": "[Status: 500 Internal Server Error]",
+    "Errors": [
+        {
+            "ErrorCode": "UK.OBIE.UnexpectedError",
+            "Message": "Internal error [Unknown Signing Algorithm]"
+        }
+    ]
+}
+
+Validate signature
+- You can validate the JWT SET signature in: https://jwt.davetonge.co.uk/
+- Copy a JWT 'set' value and paste in the web page.
+- JWKs URL endpoint to validate the signature: https://keystore.openbankingtest.org.uk/0015800001041REAAY/0015800001041REAAY.jwks
+ */
+
+def fapiInteractionId = request.getHeaders().getFirst("x-fapi-interaction-id");
+if (fapiInteractionId == null) fapiInteractionId = "No x-fapi-interaction-id"
+SCRIPT_NAME = "[SignEventsResponse] (" + fapiInteractionId + ") - "
+
+Map<String, Object> critClaims = Map.of(
+        "http://openbanking.org.uk/iat", System.currentTimeMillis() / 1000,
+        "http://openbanking.org.uk/iss", aspspOrgId,
+        "http://openbanking.org.uk/tan", "openbanking.org.uk")
+
+next.handle(context, request).thenOnResult({ response ->
+    logger.debug("{} Running...", SCRIPT_NAME)
+
+    Status status = response.getStatus()
+
+    if (status.isClientError() || status.isServerError()) {
+        return newResultPromise(response)
+    }
+
+    var responseBody = response.getEntity().getJson()
+
+    Map<String, String> sets = responseBody.sets
+
+    def slurper = new JsonSlurper()
+
+    /* No blocking call:
+     * - Compute the signature for each SET retrieved from the Test utility bank (RS) response in a loop
+     * - When all promises have been succeeded then process the result (List<Map<jti, signedJwt>)
+     */
+    when(sets.collect(entry -> {
+        Map payloadMap = slurper.parseText(entry.value)
+        return signer.sign(payloadMap, critClaims)
+                .then(signedJwt -> {
+                    return Map.entry(entry.key, signedJwt)
+                })
+    })).then(onResult -> onResult.forEach {
+        map ->  responseBody.sets[map.key] = map.value
+    }, sapiJwsSignerException -> {
+        logger.error("{} Signature fails: {}", SCRIPT_NAME, sapiJwsSignerException.getMessage())
+        response.status = Status.INTERNAL_SERVER_ERROR
+        response.entity = "{ \"error\":\"" + sapiJwsSignerException.getMessage() + "\"}"
+        return newResultPromise(response)
+    })
```

Comment:
Promises.when returns a Promise, but this is being lost. Only when this Promise has been completed do you get the final result, therefore you need to return this Promise so the framework can wait for its completion.

Again, your code only works because line 86 and line 100 are executed sequentially by the same thread, in an async situation line 86 will only run a long time after line 100 which means that the original response from the RS will already have been returned unsigned.

Javadoc for when:
> 
     * Returns a {@link Promise} which will be completed once all of the
     * provided promises have succeeded, or as soon as one of them fails.
     *
     * @param <V>
     *            The type of the tasks' result, or {@link Void} if the tasks do
     *            not return anything (i.e. they only has side-effects).
     * @param <E>
     *            The type of the exception thrown by the tasks if they fail, or
     *            {@link NeverThrowsException} if the tasks cannot fail.
     * @param promises
     *            The list of tasks to be combined.
     * @return A {@link Promise} which will be completed once all of the
     *         provided promises have succeeded, or as soon as one of them
     *         fails.
     */
    public static <V, E extends Exception> Promise<List<V>, E> when(final List<Promise<V, E>> promises) {

If you make the change that I mentioned in my previous comment, then the response handler block can itself return a Promise. Once you start using Promises then you need to keep chaining them, and putting the logic in the handlers in order to get to the final result.

The handler on line 85 will need to be responsible for producing the end Response object and returning it. All the code after line 92 needs to be deleted.